init: Require explicit -asmap filename

ryanofsky opened the pull request:
Currently, if `-asmap` is specified without a filename bitcoind tries to load `ip_asn.map` data file.

This change now requires `-asmap=ip_asn.map` or another filename to be specified explicitly.

The change is intended to make behavior of the option explicit and avoid confusion reported https://github.com/bitcoin/bitcoin/issues/33386 where documentation specifies a default file which is not actually loaded by default. It was originally implemented in
https://github.com/bitcoin/bitcoin/pull/33631#issuecomment-3410302383 and various alternatives are discussed there.

ryanofsky committed:
init: Require explicit -asmap filename

Currently, if `-asmap` is specified without a filename, bitcoind tries to load
`ip_asn.map` data file.

This change now requires `-asmap=ip_asn.map` or another filename to be
specified explicitly.

The change is intended to make behavior of the option explicit avoid confusion
reported https://github.com/bitcoin/bitcoin/issues/33386 where documentation
specifies a default file which is not actually loaded by default. It was
originally implemented in
https://github.com/bitcoin/bitcoin/pull/33631#issuecomment-3410302383 and
various alternatives are discussed there.

Co-authored-by: Fabian Jahr <fjahr@protonmail.com>

brunoerg reviewed (approved):
code review ACK cfeb160baec1369452c42d05c51f2a6af76ed077

sedited reviewed (approved):
ACK cfeb160baec1369452c42d05c51f2a6af76ed077

pablomartin4btc commented on test/functional/feature_asmap.py:
```diff
@@ -80,51 +79,48 @@ def test_asmap_with_relative_path(self):
-        shutil.copyfile(self.asmap_raw, self.default_asmap)
+        msg = f"Error: Could not find asmap file \"{self.datadir}{os.sep}\""
         for arg in ['-asmap', '-asmap=']:
             self.log.info(f'Test bitcoind {arg} (using default map file)')
```
nit:
```suggestion
            self.log.info(f'Test bitcoind {arg} (using a map file)')
```

pablomartin4btc reviewed (commented):
utACK cfeb160baec1369452c42d05c51f2a6af76ed077

If it won't be a default asmap filename, perhaps we could remove "default" from the test function names in `test/functional/feature_asmap.py`?

Left another [nit](https://github.com/bitcoin/bitcoin/pull/33770/files#r2491716003)...

ryanofsky commented:
> [P]eople were indifferent about the default path or found it nice to have. I don’t recollect anyone expressing negativity about it, as in it doesn’t make sense or it won’t be used

To be sure, I think it could make sense to have a default path that is loaded by default. I think it just to confusing to have a default path that is not loaded by default, and I think https://github.com/bitcoin/bitcoin/issues/33386 shows I'm not the only one who finds the current situation confusing.

Also if #28792 will require `-asmap=<file>` to load file data, then it would seem to make sense to also require that now, instead of providing two ways of loading file data now (`-asmap=<file>` and `-asmap`) and then taking one away later.

jurraca commented:
ACK 44717da29b626bfdcb20a32318689d74c4113b7b 

I think it's a better pattern to force users to specify the filename, but I can't tell whether users currently depend on the default.

0xB10C commented:
I load the `asmap.dat` file on by using `asmap=/absolute/path/to/asmap.dat` in my [config](https://github.com/0xB10C/peer-observer-infra-library/blob/e14fc1e29a3adbd35d5e3af73766039279fbf5c4/modules/node/node.nix#L225). Since this PR doesn't change anything for me, I'm neutral on this too.

fjahr reviewed (commented):
Approach ACK

While not everyone I asked has gotten back to me yet, almost all responders did already use an explicit file path and the few that had used the option of the default file had similar sentiment to @sipa.

The file should probably also be dropped from `files.md` then since there is nothing special about that name/location anymore, e.g. https://github.com/fjahr/bitcoin/commit/e0283bb8c133c8cae19c48c8d78cfe95dab664ee

ryanofsky reviewed (commented):
Thanks for the reviews! Made all suggested changes

<!-- begin push-4 -->
Updated 44717da29b626bfdcb20a32318689d74c4113b7b -> deac31dd779d28a8c01f1c3153650f5827cfe75d ([`pr/asmapd.3`](https://github.com/ryanofsky/bitcoin/commits/pr/asmapd.3) -> [`pr/asmapd.4`](https://github.com/ryanofsky/bitcoin/commits/pr/asmapd.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/asmapd.3..pr/asmapd.4))<!-- end --> updating files.md and adding a more specific error message for a missing filename

vostrnad commented on src/init.cpp:
```diff
@@ -532,7 +531,7 @@ void SetupServerArgs(ArgsManager& argsman, bool can_listen_ipc)
 
     argsman.AddArg("-addnode=<ip>", strprintf("Add a node to connect to and attempt to keep the connection open (see the addnode RPC help for more info). This option can be specified multiple times to add multiple nodes; connections are limited to %u at a time and are counted separately from the -maxconnections limit.", MAX_ADDNODE_CONNECTIONS), ArgsManager::ALLOW_ANY | ArgsManager::NETWORK_ONLY, ...
-    argsman.AddArg("-asmap=<file>", strprintf("Specify asn mapping used for bucketing of the peers (default: %s). Relative paths will be prefixed by the net-specific datadir location.", DEFAULT_ASMAP_FILENAME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-asmap=<file>", strprintf("Specify asn mapping used for bucketing of the peers (default: none). Relative paths will be prefixed by the net-specific datadir location."), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
I also don't like the inconsistent `(default: none)`. I think it's pretty clear that if no default is mentioned, the feature is disabled. The help text for `-i2psam` should be fixed.

ryanofsky committed:
doc: Drop (default: none) from -i2psam description

Suggested by Vojtěch Strnad (vostrnad)
https://github.com/bitcoin/bitcoin/pull/33770#discussion_r2529117675

ryanofsky reviewed (commented):
Thanks for the reviews!

<!-- begin push-5 -->
Updated deac31dd779d28a8c01f1c3153650f5827cfe75d -> 288b8c30be42f2879d1a1f3d5ec3cac13f87ace4 ([`pr/asmapd.4`](https://github.com/ryanofsky/bitcoin/commits/pr/asmapd.4) -> [`pr/asmapd.5`](https://github.com/ryanofsky/bitcoin/commits/pr/asmapd.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/asmapd.4..pr/asmapd.5))<!-- end --> making suggested documentation tweaks (since previous ACKs were stale anyway)

achow101 commented:
@fjahr DrahtBot is ignoring your ACK, is that intended? If you mean to withdraw your ACK, I think editing your comment with a strikethrough and explanation is the typical way people do that.

fjahr commented:
re-ACK 288b8c30be42f2879d1a1f3d5ec3cac13f87ace4

Only minor fixes since last review.
